### ISSUE TITLE
feature request: private user blocklists/blacklists

### ISSUE THREAD

**Issue by tracure1337** [searx/searx#2001]
please add a way to have lists of sites blocked automatically from search results.

for example I simply don't want to see results from certain "tech-sites" as they are full of ads. 

I'd consider that a killer feature.

Thanks for consideration.

**Comment by seniorm0ment** [comment]
+1 would like to see this

**Comment by HLFH** [comment]
I am not sure your example is valid regarding ads. Do you have other examples? Or can you detail more @cocoonkid or @grravity?

Ads can be blocked with an adblocker such as [uBlock Origin](https://github.com/gorhill/uBlock).
This is how people block ads automatically on the Web.
And uBlock Origin has whitelist/blacklist features.

Tech sites that have ads can be interesting as well and often are. Instead of blocking the site entirely and removing it from the search results (this is similar to @cocoonkid doing his own DMCA takedown request on search results), why not  blocking only the ads?

I am against the censorship of search results like Google is doing based on DMCA requests and its own motives.
I am in favour of the censorship of ads... with [uBlock Origin](https://github.com/gorhill/uBlock). Or using an alternative browser such as [Brave](https://brave.com/).

searx is a meta-search engine that has been built against the censorship by using various search engines.

More than a killer-feature, it will kill the essence of searx in my view, even if it is limited to one searx instance. It will bring fragmentation on the quality of meta-search results as each instance will be different with these blocklists.

The web is wide, I am not sure any kind of implementation regarding private user blocklist/blacklist is helpful. I want the best results from the best search engines on a privacy-aware meta-search engine. If some search results have been ranked at a high level, there is a reason for that. 

I think improving [meta-search](https://github.com/asciimoo/searx/issues/915) in general is more interesting.
Happy to review my opinion.

**Comment by seniorm0ment** [comment]
I do not agree with ad blocking, I agree use uBlock Origin and throw on a domain blocker like PiHole.

But as for search engine blacklists, I have heard this from a few people that use SearX here and there that they'd like to be able to blacklist certain sites.
For a very loose example, if they wanted to block a specific site that keeps showing up that they don't want in appearing anymore, one that maybe fairly consistently provides innacurate news or poor articles.
Of course this could be blocked on other levels such as domain, but this wouldn't actually filter it out from appearing in the search results.

In a result of blocking sites they have zero interest in, it would push up sites closer to view that they would have an interest in.

>The web is wide, I am not sure any kind of implementation regarding private user blocklist/blacklist is helpful. I want the best results from the best search engines on a privacy-aware meta-search engine. If some search results have been ranked at a high level, there is a reason for that.

Rankings are not always the reason of the site being good,  but the site maintaining active viewership. A lot of newsites rush out articles so they get the first click, that doesn't mean the articles they put out are good, complete, or even accurate. Now you can argue it'd be better not to block the site so it still gives a certain level of awareness to the topic  they search, but I think this should be left up to user decision whether they wish to block it or not. 
And again, this isn't limited to just news sites, it's limited to all sites. If a site constantly keeps appearing, that you have zero interest in ever visiting, then I'd be ok with blocking it to push up more preferable results and also just not have to see that site again.
I am in no way saying let servers introduce blacklists, I am suggesting let the user introduce a blacklist. They personally can choose which sites they do and do not want to see. It's not killing the web, it's just hiding the sites the users *choose* not to see.

I also agree that the quality of the results in the first place should be improved, this is in no way a replacement for that at all, but just a quality of life feature for users. Sometimes you cannot improve stuff without removing stuff. I say give the users a little freedom and let them block stuff, but also continue to improve SearX's results as well. 

On a sidenote, I am curious, does SearX do anything to combat artificially boosted results? For example, Google has been known to push a bias towards certain sites, if you enable Google's engine, does that mean those sites will start appearing closer to the top of your results? I would be highly for a way around this if it existed and currently was not in place.

**Comment by return42** [comment]
Searx is a metasearch engine, assembling the results from other search engines as they come in. It does not have an index or statistics and by this it can't rank results, like google it does.  It is just a single component to protect privacy. 

OT: To protect privacy you should also consider to use: DNS blocker, CDN blocker, Canvas blocker .. and content blocker (uBlock) ..  and drop your google, facebook ... accounts.  Drop your Win-10, OS-X ... iOS, Android .. At best don't use internet nor any connected device and buy a car from the 80's of the last century ;)

**Comment by seniorm0ment** [comment]
@return42 So essentially it does take Google's boosted results and serves them as so?

**Comment by return42** [comment]
More ore less, you can _weight_ engines and there is a simple scoring: https://github.com/asciimoo/searx/issues/1790#issuecomment-570398506

**Comment by tracure1337** [comment]
grravity perfectly explained the use case. It's not about ads actually as much it is about being able to filter out websites that have massive SEO efforts and through them simply appear so often and again and again that they clutter my results. 

I'd rather have more results of ANY OTHER but them/those. It also helps me to focus on the gems and compile a list of paywalled tech news where I rather have small blogs visible because they are from knowledgable engineers.
Especially considering that more and more content production gets automated.

I already know I will not click them so why not have a way to automate their appearance/disappearance.

Again this blocklist would be private so it only pertains to the user himself.

Thank you for consideration.

**Comment by tracure1337** [comment]
Regarding privacy you have great suggestions and I am happy to tell that I have implemented nearly all of them.

**Comment by return42** [comment]
> Regarding privacy you have great suggestions and I am happy to tell that I have implemented nearly all of them.

and the car? :)

> It's not about ads actually as much it is about being able to filter out websites that have massive SEO

To be clear, I do not vote against a _blocklist_ at all. I just wanted to clarify, that searx does not have in index. A filter could be implemented.

**Comment by tracure1337** [comment]
I am saving for the Delorian 🤣 

Looking forward to a possible implementation and thanks again to everyone who has taken the time to reply!

**Comment by mokulus** [comment]
@dalf @asciimoo 
I have a plan for this:
- add blacklist file path in settings and load it
- compile the the regexes and save them in settings (might not be the best place), so we don't read the file every time
- filter results in Search.search, just before returning by removing results which match any regex in blacklist, using result['url']

You could also make transform regexes similar to: https://github.com/asciimoo/searx/issues/2080, with simple sed-like syntax
/^http[^s]/https/
|//twitter.com/|//nitter.net/
Which could be in a "transformlist"

You could parse it, and use https://docs.python.org/3/library/re.html#re.sub for that.

I think I could do it, because it seems pretty easy, it's basically using python re lib. But I'm unfamiliar with the project. Give me a green light if this seems ok.

I'm unfamiliar with the plugins but I think it could be better to just have a plugin type that would process results, changing whatever it want in the dict.

**Comment by asciimoo** [comment]
@mokulus This sounds like a good plan.

> compile the the regexes and save them in settings (might not be the best place), so we don't read the file every time

I don't think reading/compiling on every startup would be a big overhead. I'd specify only the rule file in settings.yml and read it on every startup in initialization time.

> I think I could do it, because it seems pretty easy, it's basically using python re lib. But I'm unfamiliar with the project. Give me a green light if this seems ok.

It would be great, but I'd like to hear the opinion of the other maintainers/core devs. @dalf @kvch @HLFH @MarcAbonce  any remarks?

> I'm unfamiliar with the plugins but I think it could be better to just have a plugin type that would process results, changing whatever it want in the dict.

I wouldn't go with a plugin, especially because it would be good to display somewhere on the UI these rules to the users to avoid confusion. Even if it isn't a plugin we can add a settings option to turn it off if someone doesn't want to use it.

**Comment by mokulus** [comment]
I meant saving in settings as saving in settings dictionary in code as objects. Not sure how else to pass them to Search.

**Comment by dalf** [comment]
To be sure:
* the purpose is to remove some URL from the results.
* by _certain "tech-sites"_ (in the OP), I understand domain name: in this case, we don't need regex, just check if ```result['parsed_url'].hostname``` is inside a list ?
* but regex could be useful to remove the spam accounts ( ```//example.com/spamcount/.*``` ? )

About the list: who decide which URL are blocked ?
1. is it a choice of the user ? if yes, there is a "hidden these websites" textarea in the user preferences.
   * no maintenance
   * difficult / impossible to implement without using an user session (no-go feature). Without sessions, a huge list will be sent over and over using the cookies, even if [HPACK](https://blog.cloudflare.com/hpack-the-silent-killer-feature-of-http-2/) can help we can't assume all connections will use HTTP/2.
1. or is it community based effort to produce a different list(s) (one list or multiple list: tech ads, spam, ...). If yes, there is a "hidden the website from this list" in the user preferences.
   * easy to implement
   * maintenance / discussion about which website should be included or not.
1. or something else.

@mokulus I understand you want to implement the second case, a static list ? If yes, I would put this list(s) in the searx/data folder, where there are already the https_everywhere rules.

About the format: , a json file ? Feel free to do the way you want (I agree with @asciimoo , compile the regex on startup won't be a big overhead).
* in JSON, it can be:
```json
[
	{ "host": "example.org" },
	{ "regex": "//example.com/spamaccount/" }, 
]
```
* [easylist](https://github.com/easylist/easylist/tree/master/easylist) has one text file per list (one line, one rule). Simple, easy to edit and to parse.
* https_everywhere one xml file per company ( google.xml, github.xml, etc...): easy to edit in this case, and then there is a script to compile all the rules in one big files (but I don't think this would be useful here).

For now we can forget about https://github.com/asciimoo/searx/issues/2080 : if I sum up the purpose is to trigger "on_result" plugin only for some URL. But it is not clear for me how to do it (technically yes, but I'm not sure about the API).

> I'm unfamiliar with the plugins but I think it could be better to just have a plugin type that would process results

@asciimoo do you think it should be an external plugin ? I think about the list content management (linked issue: https://github.com/asciimoo/searx/issues/2052 and this ["brainstorming"](https://github.com/asciimoo/searx/wiki/Brainstorming:-IDE-&-database#searx-engine-database) ).

**Comment by seniorm0ment** [comment]
> 1. is it a choice of the user ? if yes, there is a "hidden these websites" textarea in the user preferences

Personally, this is what I'm hoping for, a user controllable blocklist. Now it would be nice if this could be exported/imported with ease.

**Comment by asciimoo** [comment]
> Personally, this is what I'm hoping for, a user controllable blocklist.

It can be problematic to save big lists, because we don't persist any user settings, we just store them in cookies.

**Comment by mokulus** [comment]
@grravity @cocoonkid 
If you use uBlock orgin you can add a custom filter to block results(click on uBlockOrigin icon, click fourth button, click my filters) , using `xpath`, for example:
```
searx.ninja##:xpath(//div[contains(@class, 'result')]/div[contains(@class, 'external-link')][contains(., 'reddit')]/..)
```
Consult https://en.wikipedia.org/wiki/XPath for information on how it works exaclty. The interesting part is `contains(., 'reddit')`. Its a simple match, regexes are not supported.

**Comment by MarcAbonce** [comment]
Related issue: #407.

I think it could work either as a plugin or as a separate feature, but I think it could be easier to make it a plugin.

It's worth pointing out that there's also the semi-related #1878 which filters results by querying a database or an external API. And also, I've wanted to implement a separate unrelated blacklist for my instance, but before doing so I wanted to see how any of these issues/PRs turns out so I can base my implementation on whatever gets merged into master. I'm sure #1878 can also be easily rewritten, so don't worry about binding yourself to that implementation, I just want to point out that different people want to use different blacklists so we should take that into account.

My only worry is that we could potentially end up with many inconsistent implementations of blacklists spread around the application. So in my opinion, we need to take into account that people may want to add more than one blacklist and that each one of those could work differently (with regex, or a db or in my case a file of md5 hashes) but all of them should look the same in the UI.

If they're implemented as plugins, I think we could use the plugin's `preference_section` attribute to add a new section for result filters/blacklists so we can show them together in the preferences view, but I'm not very sure how that attribute actually works.

**Comment by mbirth** [comment]
Just for the record, out of my head I can come up with 2 sites I definitely don't want to see in my search results yet get them over and over:

* w3schools
* Pinterest

So I guess the blacklist for each user won't be THAT long. 10 should be more than enough to filter out the extreme cases, 50 would be luxurious. I think this can still fit in a cookie. We're talking about SEO spam sites here, not filtering out any page I don't like.

**Comment by gripped** [comment]
I came here after looking in the settings to see if there was a way I could blacklist Pinterest.  (I host my own remote instance, used only by me).
Then thought I'd come here, look to see if such a feature had been requested before and if not request it. Obviously it has.
I was amused to see the last post is also a hater of Pinterest.

Anyway a simple blacklist saying I never want to see results from:
a.com
b.com
c.co.uk etc.
would be great.

But I appreciate it may not actually be simple to implement, as I imagine more people use searx through public, or at least shared, instances ? And you have to account for that.

Anyhow while I'm here thanks. I use (my) searx for about 90% of my searches. It's great.

**Comment by katacarbix** [comment]
I miss this feature from Google search that it had for a short time back in 2011. I currently would like to block Pinterest from results but I'm sure I could think of more. I hope this gets implemented soon, either as a plugin or as a core feature.

**Comment by ralphrmartin** [comment]
It seems some people have confused blocking sites (as in adblocking) with filtering out unwanted results in the search results.
There seem to be ever more fairly useless "tech sites" for example that just repeat other information from elsewhere, or are written by non-experts. I would like very much to be able to stop results from including these sites.
I believe this has to be done on a personal level. What is one person's junk may be found to be interesting by someone else.
Above, it was suggested that 50 such sites might be a maximum, but I could easily see this running to hundreds of sites if not more.

However there are 2 use cases to consider - public searx instances, and private ones.
Private ones are easier to deal with, and the "kill list" could just be a simple list of sites stored as a preference file on the server (one per user, if there are multiple users, perhaps).
Public servers clearly cannot and do not want to store such information, but sending such a long personal list for each query does not make sense either.

Perhaps the way forward is to make this a feature for private servers only - which could encourage more take-up of private instances of searx (and reduce load on public servers as a by-product?).

**Comment by xeekdev** [comment]
Has anyone found a way to implement this? I made a plugin that returns False in on_result() for domains I don't want but that does not seem to work. What is the proper way to have on_result() discard a result?

**Comment by jt196** [comment]
Came here after seeing a tech site indexer for the nth time in my search results. I don't want to look at a page that indexes Stackoverflow, I'd rather go direct. My first thought was to look in the config. I'll use @mokulus' method for the time being, thanks.

**Comment by astier** [comment]
I would like to block pinterest and slant. Slant always shows me totally useless computer/AI-generated comparisons between programs. There are some other like libhunt and saashub. An example of such useless result would be:

https://www.saashub.com/compare-swisscows-vs-searx

I can't remember a single time when this sites were useful. Actually they waste my time because somehow they still manage to trick me once in a while to click on their link because the title is intriguing but I don't realize its from them. For me this sites only produce spam in my search-results.

NOTE: I'm not talking about adblockers like ublock. I already have ublock installed. Or if I can use ublock to somehow make results from this sites disappear on my private searx-instance then please explain.

**Comment by mbirth** [comment]
Just for the record: There's also [uBlacklist](https://github.com/iorate/uBlacklist).

However, that plugin removes results from the page. So e.g. with 2 blocked sites instead of 10 results on a page, you'd get 8. So a native blocking feature would be better.

**Comment by jt196** [comment]
@astier - scroll up, @mokulus has given [a uBlock method](https://github.com/searx/searx/issues/2001#issuecomment-665107289), works fine. 

`searx.example.com##:xpath(//div[contains(@class, 'result')]/div[contains(@class, 'external-link')][contains(., 'codegrepper')]/..)` 

...in the "Your Filters" section of the uBlock settings to block codegrepper.

Obvs, you'd need to configure this per browser, but it's exportable, and I guess the uBlock settings sync across browsers if you're signed in.

**Comment by ak3io** [comment]
> Just for the record: There's also [uBlacklist](https://github.com/iorate/uBlacklist).
> 
> However, that plugin removes results from the page. So e.g. with 2 blocked sites instead of 10 results on a page, you'd get 8. So a native blocking feature would be better.

It would also be better to have it integrated in the searx preferences because some of us, in our quest to have a better web, don't use browsers that have a plugin ecosystem :)

Big +1 for this feature request! I have several sites I block with Pihole and I'd like to exclude them from results altogether. I'm running my own instance on my home network, if that helps anyone understand the use case.

**Comment by l3gacyb3ta** [comment]
Personally, I have a group of folx I would love to build a small-internet search engine for, by excluding the top 1000 sites on the internet. This would be perfect, because I don't want to have to rewrite searx myself.

**Comment by reeseovine** [comment]
I've actually been using [LetsBlockIt](https://letsblock.it/filters/search-results) lately to block certain websites from my SearX results. It generates a uBlock Origin filter list so it's not a perfect solution for all use cases but it's something while this issue gets worked on.

**Comment by ryliejamesthomas** [comment]
Similar to others' UBlock method, I add rules to Firefox's userContent.css to filter out results.

I'd love to be able to build the list into a SearX instance.

**Comment by ralphrmartin** [comment]
Using LetsBlockIt and UBlock Origin doesn't seem to work for me. UBlock Origin just gives a network error every time I try to fetch my list from LetsBlockIt.

**Comment by jt196** [comment]
SearxNG has this [here](https://github.com/searxng/searxng/discussions/970)

**Comment by ralphrmartin** [comment]
> SearxNG has this [here](https://github.com/searxng/searxng/discussions/970)

Thanks. Any pointers on how to install and use this? I guess that would help a lot of people reading this.

**Comment by jt196** [comment]
If you're running a SearxNG instance, you just add that code to the config yml:
```yml
hostname_replace:
  # pages which reflect nothing more than bad translations of other pages
  'com\.de': false
  'qastack\.com\.de': false
  'im-coder\.com': false
  'code\.i-harness\.com': false
```

**Comment by ralphrmartin** [comment]
Thanks. For those looking for a recipe: 
Edit /etc/searxng/settings.yml so that it includes

```
enabled_plugins:
   - 'Hostname replace'  # see hostname_replace configuration below
hostname_replace:
  'im-coder\.com': false  # one or more lines like this, one for each site you don't want to appear in the search results 

```

**Comment by jt196** [comment]
Thanks @ralphrmartin was on mobile so not chance to check my SearxNG yml

**Comment by masterofobzene** [comment]
I opened this request for SEARXNG also, its confusing don't know who is the main branch anymore.  
Anyway here is the issue I rised there.

> Implement an option to remove certain results from searches by using phrases or keywords.
> I'm actually using the [ublacklist extension that I modified to work with SEARX (firefox only)](https://github.com/masterofobzene/SearchBlocklist), and I know the "-" option in the search field to remove certain words but doing this takes more time when searching than simply searching for the intended words/phrase. Also ublacklist needs specific URLs to work so I would have to repeat the blocking on every bad result (that are probably auto-generated using search keywords so this wold be infinite).
> 
> WHY?
> I'm really... really... REALLY tired of "toplists"; results like "5 Ways to....", "10 best softwares to...", "3 things..." etc etc.
> Recently those websites started to pop-up like a virus on the net.
> 
> IDEA(?): An option that auto-concatenates certain "negative words" (i.e "-10 best") that the user can configure via settings so when searching, those options will be automatically appended to the search i.e "cut mp3 -"10 best" "

**Comment by craftyguy** [comment]
searxng has been working fine for me without any janky browser plugins, with configuration like this (as @jt196 pointed out earlier):

```
hostname_replace:
  '(.*\.)?amazon\.com': false
  '(.*\.)?github\.(.*\.)?': false
  '(.*\.)?microsoft\.com': false
  '(.*\.)?amzn\.to': false
```

**Comment by masterofobzene** [comment]
> searxng has been working fine for me without any janky browser plugins, with configuration like this (as @jt196 pointed out earlier):
> 
> ```
> hostname_replace:
>   '(.*\.)?amazon\.com': false
>   '(.*\.)?github\.(.*\.)?': false
>   '(.*\.)?microsoft\.com': false
>   '(.*\.)?amzn\.to': false
> ```

Great, except:

A- Nobody said Searx works bad.
B- Blocking domains is NOT what I asked above.

> Implement an option to remove certain results from searches by using phrases or keywords.

> Also ublacklist needs specific URLs to work so I would have to repeat the blocking on every bad result (**that are probably auto-generated using search keywords so this would be infinite**). 

This means, "bad results" have different domains. Keyword blocking would be much more efficient than domain blocking.
Imagine yourself searching for something and then having to block all random-spam-blog sites that pop up in searches one by one. Go ahead and do the experience, try searching for some software that does something in particular, like "cut mp3". You will get 30% toplist sites and 50% "online" based """services""".

My idea is to set predefined "negative words" like you would normally use when searching, but to make it easier, you could just configure them to append automatically to your search.  
For example: to filter the damn toplist and online service sites, I could configure the words: "best", "ways", "online" in the <<"theorical">> settings.
Then search for <<"cut mp3">> 

and see this in the search textbox:
cut mp3 -"best" -"ways" -"online"

**Comment by slashformotion** [comment]
is this issue still alive? This is a feature I would love. 

I use searx a lot to search programming related subjects and I would love to remove "code greppers" sites that just scrape content from somewhere else:
example: I am searching information that related to the golang module "lipgloss" and there is this site https://golangexample.com/style-definitions-for-nice-terminal-layouts-built-with-golang/ that just copied the README of the project.

**Comment by masterofobzene** [comment]
> is this issue still alive? This is a feature I would love.
> 
> I use searx a lot to search programming related subjects and I would love to remove "code greppers" sites that just scrape content from somewhere else: example: I am searching information that related to the golang module "lipgloss" and there is this site https://golangexample.com/style-definitions-for-nice-terminal-layouts-built-with-golang/ that just copied the README of the project.


See https://github.com/searxng/searxng/issues/1258

**Comment by sjehuda** [comment]
I've wrote a userscript that filters results by hostnames
https://openuserjs.org/scripts/sjehuda/SearX_Results_Filter

You need to add the URL of your favorite SearxNG instance
If you want to add/ remove hostname, simply edit the domains list in this script.

**Comment by ScaredyCat** [comment]
> Thanks. For those looking for a recipe: Edit /etc/searxng/settings.yml so that it includes
> 
> ```
> enabled_plugins:
>    - 'Hostname replace'  # see hostname_replace configuration below
> hostname_replace:
>   'im-coder\.com': false  # one or more lines like this, one for each site you don't want to appear in the search results 
> ```

This doesn't seem to work with the image search (which is where I don't want pinterst results the most)
